File: ThalesCore/Cryptography/HexKey.vb
''
'' This program is free software; you can redistribute it and/or modify
'' it under the terms of the GNU General Public License as published by
'' the Free Software Foundation; either version 2 of the License, or
'' (at your option) any later version.
''
'' This program is distributed in the hope that it will be useful,
'' but WITHOUT ANY WARRANTY; without even the implied warranty of
'' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'' GNU General Public License for more details.
''
'' You should have received a copy of the GNU General Public License
'' along with this program; if not, write to the Free Software
'' Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
'' 

Namespace Cryptography

    ''' <summary>
    ''' Represents a hexadecimal single, double or triple length key.
    ''' </summary>
    ''' <remarks>
    ''' Objects of this class are used as placeholders for key data. 
    ''' </remarks>
    Public Class HexKey

        ''' <summary>
        ''' Key length enumeration.
        ''' </summary>
        ''' <remarks>
        ''' This is an enumeration that defines the length of a hexadecimal key.
        ''' </remarks>
        Public Enum KeyLength
            ''' <summary>
            ''' Single length keys.
            ''' </summary>
            ''' <remarks>
            ''' Defines a single length hexadecimal key (16 digits).
            ''' </remarks>
            SingleLength = 0

            ''' <summary>
            ''' Double length keys.
            ''' </summary>
            ''' <remarks>
            ''' Defines a double length hexadecimal key (32 digits).
            ''' </remarks>
            DoubleLength = 1

            ''' <summary>
            ''' Triple length keys.
            ''' </summary>
            ''' <remarks>
            ''' Defines a triple length hexadecimal key (48 digits).
            ''' </remarks>
            TripleLength = 2
        End Enum

        Private _partA As String
        Private _partB As String
        Private _partC As String
        Private _keyLength As KeyLength

        ''' <summary>
        ''' First part of the key.
        ''' </summary>
        ''' <remarks>
        ''' Returns or sets the first part of the represented key.
        ''' </remarks>
        Public Property PartA() As String
            Get
                Return _partA
            End Get
            Set(ByVal Value As String)
                _partA = Value
            End Set
        End Property

        ''' <summary>
        ''' Second part of the key.
        ''' </summary>
        ''' <remarks>
        ''' Returns or sets the second part of the represented key.
        ''' </remarks>
        Public Property PartB() As String
            Get
                Return _partB
            End Get
            Set(ByVal Value As String)
                _partB = Value
            End Set
        End Property

        ''' <summary>
        ''' Third part of the key.
        ''' </summary>
        ''' <remarks>
        ''' Returns or sets the third part of the represented key.
        ''' </remarks>
        Public Property PartC() As String
            Get
                Return _partC
            End Get
            Set(ByVal Value As String)
                _partC = Value
            End Set
        End Property

        ''' <summary>
        ''' Key length.
        ''' </summary>
        ''' <remarks>
        ''' Returns or sets the key length.
        ''' </remarks>
        Public Property KeyLen() As KeyLength
            Get
                Return _keyLength
            End Get
            Set(ByVal Value As KeyLength)
                _keyLength = Value
            End Set
        End Property

        ''' <summary>
        ''' Key constructor.
        ''' </summary>
        ''' <remarks>
        ''' The constructor sets internal variables according to the length of the 
        ''' supplied key. Use the <see cref="Cryptography.HexKey.KeyLen"/> property
        ''' to determine the length of the key instead of the <see cref="Cryptography.HexKey.PartA"/>,
        '''  <see cref="Cryptography.HexKey.PartB"/> and <see cref="Cryptography.HexKey.PartC"/>
        ''' properties. Internally, the key is always represented as a triple length key according to the
        ''' following rule:
        ''' <P>If the key data is 16 digits long, all key parts are set to this value.</P>
        ''' <P>If the key data is 32 digits long, <see cref="Cryptography.HexKey.PartA"/> and
        ''' <see cref="Cryptography.HexKey.PartC"/> are both set to the first 16 digits, and
        ''' <see cref="Cryptography.HexKey.PartB"/> is set to the last 16 digits.</P>
        ''' <P>If the key data is 48 digits long, all key parts are distributed accordingly.</P>
        ''' </remarks>
        Public Sub New(ByVal key As String)

            If key Is Nothing OrElse key = "" Then
                Throw New Exceptions.XInvalidKey("Invalid key data: " + key)
            End If

            If key.Length = 17 OrElse key.Length = 33 OrElse key.Length = 49 Then
                Dim scheme As ThalesSim.Core.KeySchemeTable.KeyScheme = ThalesSim.Core.KeySchemeTable.GetKeySchemeFromValue(key.Substring(0, 1))
                If scheme <> Core.KeySchemeTable.KeyScheme.DoubleLengthKeyAnsi AndAlso _
                   scheme <> Core.KeySchemeTable.KeyScheme.DoubleLengthKeyVariant AndAlso _
                   scheme <> Core.KeySchemeTable.KeyScheme.SingleDESKey AndAlso _
                   scheme <> Core.KeySchemeTable.KeyScheme.TripleLengthKeyAnsi AndAlso _
                   scheme <> Core.KeySchemeTable.KeyScheme.TripleLengthKeyVariant Then
                    Throw New Exceptions.XInvalidKeyScheme("Invalid key scheme " + key.Substring(0, 1))
                Else
                    key = key.Substring(1)
                End If
            End If

            If Core.Utility.IsHexString(key) = False Then
                Throw New Exceptions.XInvalidKey("Invalid key data: " + key)
            End If

            If key.Length = 16 Then
                _partA = key
                _partB = key
                _partC = key
                _keyLength = KeyLength.SingleLength
            ElseIf key.Length = 32 Then
                _partA = key.Substring(0, 16)
                _partB = key.Substring(16)
                _partC = _partA
                _keyLength = KeyLength.DoubleLength
            ElseIf key.Length = 48 Then
                _partA = key.Substring(0, 16)
                _partB = key.Substring(16, 16)
                _partC = key.Substring(32)
                _keyLength = KeyLength.TripleLength
            Else
                Throw New Exceptions.XInvalidKey("Invalid key length: " + key)
            End If

        End Sub

    End Class

End Namespace

File: ThalesSim.Core/Cryptography/HexKey.cs
/*
 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

using System;
using ThalesSim.Core.Cryptography.DES;
using ThalesSim.Core.Utility;

namespace ThalesSim.Core.Cryptography
{
    /// <summary>
    /// This class represents a hexadecimal key used to encrypt/decrypt data.
    /// </summary>
    public class HexKey
    {
        /// <summary>
        /// Get/set the key scheme.
        /// </summary>
        public KeyScheme Scheme { get; private set; }

        /// <summary>
        /// Get/set the key length.
        /// </summary>
        public KeyLength Length { get; private set; }

        /// <summary>
        /// Get the represented by this instance.
        /// </summary>
        public string Key { get; private set; }

        /// <summary>
        /// Get the first key part.
        /// </summary>
        public string PartA
        {
            get { return Key.Substring(0, 16); }
        }

        /// <summary>
        /// Get the second key part.
        /// </summary>
        public string PartB
        {
            get { return Length != KeyLength.SingleLength ? Key.Substring(16, 16) : Key; }
            private set { Key = Length == KeyLength.DoubleLength ? PartA + value : PartA + value + PartC; }
        }

        /// <summary>
        /// Get the third key part.
        /// </summary>
        public string PartC
        {
            get { return Length == KeyLength.TripleLength ? Key.Substring(32, 16) : Key.Substring(0, 16); }
        }

        /// <summary>
        /// Creates a new instance of this class from a hexadecimal value.
        /// </summary>
        /// <param name="text"></param>
        public HexKey(string text)
        {
            if (string.IsNullOrEmpty(text))
            {
                throw new InvalidCastException("Empty hex key not allowed");
            }

            Scheme = KeyScheme.Unspecified;
            text = text.ToUpper();

            if (text.StartsWithKeyScheme())
            {
                Scheme = text.Substring(0, 1).GetKeyScheme();
                text = text.Substring(1);
            }

            if (!text.IsHex())
            {
                throw new InvalidCastException(string.Format("Key {0} is not hex", text));
            }

            switch (text.Length)
            {
                case 16:
                    if (Scheme == KeyScheme.Unspecified)
                    {
                        Scheme = KeyScheme.SingleLengthKey;
                    }
                    Length = KeyLength.SingleLength;
                    break;
                case 32:
                    if (Scheme == KeyScheme.Unspecified)
                    {
                        Scheme = KeyScheme.DoubleLengthKeyAnsi;
                    }
                    Length = KeyLength.DoubleLength;
                    break;
                case 48:
                    if (Scheme == KeyScheme.Unspecified)
                    {
                        Scheme = KeyScheme.TripleLengthKeyAnsi;
                    }
                    Length = KeyLength.TripleLength;
                    break;
                default:
                    throw new InvalidCastException(string.Format("Invalid value {0} for key", text));
            }

            Key = text;
        }

        /// <summary>
        /// Returns a string representation of this instance.
        /// </summary>
        /// <returns>String representing this instance.</returns>
        public override string ToString()
        {
            return Scheme.GetKeySchemeChar().Replace("Z","") + Key;
        }

        /// <summary>
        /// Encrypt data under this key.
        /// </summary>
        /// <param name="data">16, 32 or 48 hex data to encrypt.</param>
        /// <returns>Encrypted data.</returns>
        public string Encrypt (string data)
        {
            ValidateEncryptDecryptData(data);

            if (data.Length == 16)
            {
                return TripleDes.TripleDesEncrypt(PartA, PartB, PartC, data);
            }

            if (data.Length == 32)
            {
                return TripleDes.TripleDesEncrypt(PartA, PartB, PartC, data.Substring(0, 16)) +
                       TripleDes.TripleDesEncrypt(PartA, PartB, PartC, data.Substring(16, 16));
            }

            return TripleDes.TripleDesEncrypt(PartA, PartB, PartC, data.Substring(0, 16)) +
                   TripleDes.TripleDesEncrypt(PartA, PartB, PartC, data.Substring(16, 16)) +
                   TripleDes.TripleDesEncrypt(PartA, PartB, PartC, data.Substring(32, 16));
        }

        /// <summary>
        /// Decrypt data using this key.
        /// </summary>
        /// <param name="data">16, 32 or 48 hex data to decrypt.</param>
        /// <returns>Decrypted data.</returns>
        public string Decrypt(string data)
        {
            ValidateEncryptDecryptData(data);

            if (data.Length == 16)
            {
                return TripleDes.TripleDesDecrypt(PartA, PartB, PartC, data);
            }

            if (data.Length == 32)
            {
                return TripleDes.TripleDesDecrypt(PartA, PartB, PartC, data.Substring(0, 16)) +
                       TripleDes.TripleDesDecrypt(PartA, PartB, PartC, data.Substring(16, 16));
            }

            return TripleDes.TripleDesDecrypt(PartA, PartB, PartC, data.Substring(0, 16)) +
                   TripleDes.TripleDesDecrypt(PartA, PartB, PartC, data.Substring(16, 16)) +
                   TripleDes.TripleDesDecrypt(PartA, PartB, PartC, data.Substring(32, 16));
        }

        /// <summary>
        /// Encrypt data under this key using CBC mode.
        /// </summary>
        /// <param name="data">16, 32 or 48 hex data to encrypt.</param>
        /// <returns>Encrypted data.</returns>
        public string EncryptCbc (string data)
        {
            ValidateEncryptDecryptData(data);

            if (data.Length == 16)
            {
                return TripleDes.TripleDesEncrypt(PartA, PartB, PartC, data);
            }

            if (data.Length == 32)
            {
                var resulta1 = TripleDes.TripleDesEncrypt(PartA, PartB, PartC, data.Substring(0, 16));
                return resulta1 + TripleDes.TripleDesEncrypt(PartA, PartB, PartC, data.Substring(16, 16).XorHex(resulta1));
            }

            var resulta2 = TripleDes.TripleDesEncrypt(PartA, PartB, PartC, data.Substring(0, 16));
            var resultb = TripleDes.TripleDesEncrypt(PartA, PartB, PartC, data.Substring(16, 16).XorHex(resulta2));
            return resulta2 + resultb + TripleDes.TripleDesEncrypt(PartA, PartB, PartC, data.Substring(32, 16).XorHex(resulta2));
        }

        /// <summary>
        /// Decrypt data using CBC mode.
        /// </summary>
        /// <param name="data">16, 32 or 48 hex data to decrypt.</param>
        /// <returns>Decrypted data.</returns>
        public string DecryptCbc(string data)
        {
            ValidateEncryptDecryptData(data);

            if (data.Length == 16)
            {
                return TripleDes.TripleDesDecrypt(PartA, PartB, PartC, data);
            }

            if (data.Length == 32)
            {
                var resulta1 = TripleDes.TripleDesDecrypt(PartA, PartB, PartC, data.Substring(0, 16));
                return resulta1 + TripleDes.TripleDesDecrypt(PartA, PartB, PartC, data.Substring(16, 16)).XorHex(data.Substring(0, 16));
            }

            var resulta2 = TripleDes.TripleDesDecrypt(PartA, PartB, PartC, data.Substring(0, 16));
            var resultb = TripleDes.TripleDesDecrypt(PartA, PartB, PartC, data.Substring(16, 16).XorHex(data.Substring(0, 16)));
            return resulta2 + resultb + TripleDes.TripleDesDecrypt(PartA, PartB, PartC, data.Substring(32, 16).XorHex(data.Substring(16, 16)));
        }

        /// <summary>
        /// Encrypt data under this key using the variant mode.
        /// </summary>
        /// <param name="data">16, 32 or 48 hex data to encrypt.</param>
        /// <returns>Encrypted data.</returns>
        public string EncryptVariant (string data)
        {
            ValidateEncryptDecryptData(data);

            if (data.Length == 16)
            {
                throw new InvalidOperationException("Cannot use 8-byte data with EncryptVariant");
            }

            if (data.Length == 32)
            {
                var orgKeyPartB = PartB;
                PartB = PartB.XorHex(LMK.Variants.GetDoubleLengthVariant(1).PadRight(16, '0'));
                var result1 = Encrypt(data.Substring(0, 16));
                PartB = orgKeyPartB.XorHex(LMK.Variants.GetDoubleLengthVariant(2).PadRight(16, '0'));
                var result2 = Encrypt(data.Substring(16, 16));
                PartB = orgKeyPartB;
                return result1 + result2;
            }
            else
            {
                var orgKeyPartB = PartB;
                PartB = PartB.XorHex(LMK.Variants.GetTripleLengthVariant(1).PadRight(16, '0'));
                var result1 = Encrypt(data.Substring(0, 16));
                PartB = orgKeyPartB.XorHex(LMK.Variants.GetTripleLengthVariant(2).PadRight(16, '0'));
                var result2 = Encrypt(data.Substring(16, 16));
                PartB = orgKeyPartB.XorHex(LMK.Variants.GetTripleLengthVariant(3).PadRight(16, '0'));
                var result3 = Encrypt(data.Substring(32, 16));
                PartB = orgKeyPartB;
                return result1 + result2 + result3;
            }
        }

        /// <summary>
        /// Decrypt data using the variant mode.
        /// </summary>
        /// <param name="data">16, 32 or 48 hex data to decrypt.</param>
        /// <returns>Decrypted data.</returns>
        public string DecryptVariant (string data)
        {
            ValidateEncryptDecryptData(data);
            
            if (data.Length == 16)
            {
                throw new InvalidOperationException("Cannot use 8-byte data with DecryptVariant");
            }

            if (data.Length == 32)
            {
                var orgKeyPartB = PartB;
                PartB = PartB.XorHex(LMK.Variants.GetDoubleLengthVariant(1).PadRight(16, '0'));
                var result1 = Decrypt(data.Substring(0, 16));
                PartB = orgKeyPartB.XorHex(LMK.Variants.GetDoubleLengthVariant(2).PadRight(16, '0'));
                var result2 = Decrypt(data.Substring(16, 16));
                PartB = orgKeyPartB;
                return result1 + result2;
            }
            else
            {
                var orgKeyPartB = PartB;
                PartB = PartB.XorHex(LMK.Variants.GetTripleLengthVariant(1).PadRight(16, '0'));
                var result1 = Decrypt(data.Substring(0, 16));
                PartB = orgKeyPartB.XorHex(LMK.Variants.GetTripleLengthVariant(2).PadRight(16, '0'));
                var result2 = Decrypt(data.Substring(16, 16));
                PartB = orgKeyPartB.XorHex(LMK.Variants.GetTripleLengthVariant(3).PadRight(16, '0'));
                var result3 = Decrypt(data.Substring(32, 16));
                PartB = orgKeyPartB;
                return result1 + result2 + result3;
            }
        }

        private void ValidateEncryptDecryptData (string data)
        {
            if (string.IsNullOrEmpty(data))
            {
                throw new InvalidOperationException("Cannot use empty data with encypt/decrypt");
            }

            if (data.Length != 16 && data.Length != 32 && data.Length != 48)
            {
                throw new InvalidOperationException(string.Format("Data to encrypt/decrypt [{0}] has invalid length", data));
            }
        }
    }
}
